pkg/plugins/pg/destination/destination.go

```diff
@@ -141,24 +205,72 @@ func (s *Destination) Write(ctx context.Context, r record.Record) (record.Positi
 	// we have to manually append a semi colon to the upsert sql;
 	upsertQuery += ";"
 
-	// prepare SQL to insert cols and args into the appropriate table.
-	// suffix sql and args for upsert behavior.
-	query, args, err := psql.Insert(table).
+	colArgs, valArgs := formatColumnsAndValues(key, payload)
+
+	query, args, err := psql.
+		Insert(tableName).
 		Columns(colArgs...).
 		Values(valArgs...).
 		SuffixExpr(sq.Expr(upsertQuery)).
 		ToSql()
 	if err != nil {
-		return s.Position, cerrors.Errorf("error formatting query: %w", err)
+		return "", nil, fmt.Errorf("error formatting query: %w", err)
 	}
 
-	// attempt to run the query
-	_, err = s.db.Exec(query, args...)
-	if err != nil {
-		// return current position that hasn't been updated and the error
-		return s.Position, cerrors.Errorf("insert exec failed: %w", err)
+	return query, args, nil
+}
+
+// formatColumnsAndValues turns the key and payload into a slice of ordered
+// columns and values for upserting into Postgres.
+func formatColumnsAndValues(key, payload sdk.StructuredData) ([]string, []interface{}) {
+	var colArgs []string
+	var valArgs []interface{}
+
+	// range over both the key and payload values in order to format the
+	// query for args and values in proper order
+	for key, val := range key {
+		colArgs = append(colArgs, key)
+		valArgs = append(valArgs, val)
+		delete(payload, key) // NB: Delete Key from payload arguments
+	}
+
+	for field, value := range payload {
+		colArgs = append(colArgs, field)
+		valArgs = append(valArgs, value)
+	}
+
+	if len(colArgs) != len(valArgs) {
+		panic("how should we handle this case?")
+	}
+
+	return colArgs, valArgs
+}
+
+// return either the records metadata value for table or the default configured
+// value for table. Otherwise it will error since we require some table to be
+// set to write into.
+func (d *Adapter) getTableName(metadata map[string]string) (string, error) {
+	tableName, ok := metadata["table"]
+	if !ok {
+		if d.config.tableName == "" {
+			// NB: Should we set a default here or error?
```

Comment:
IMHO, the way you have it here (error) is better, since I don't think there's a reasonable default. I believe that every user would like to see the data in a particular table, so sending it to a default one could only mean more work (migrating the data from the default table to the desired one).